packages/next-on-pages/src/buildApplication/generateFunctionsMap.ts

```diff
@@ -698,3 +713,77 @@ async function copyWasmFiles(
 		await copyFile(originalFileLocation, newLocation);
 	}
 }
+
+/**
+ * Map that maps a chunk number to a set of chunk consumers (by consumers we mean function files, and those
+ * are saved/referenced via their filepaths)
+ */
+type ChunksConsumersInfo = Map<number, ChunkConsumers>;
+type ChunkConsumers = Set<string>;
+
+async function collectChunksConsumersInfoRecursively(
+	setup: ProcessingSetup,
+	dir: string
+): Promise<ChunksConsumersInfo> {
+	const chunksConsumersInfos: ChunksConsumersInfo = new Map();
+
+	const files = await readdir(dir);
+
+	for (const file of files) {
+		const filepath = join(dir, file);
+		if (await validateDir(filepath)) {
+			if (!file.endsWith('.rsc.func')) {
+				const dirResultsPromise = file.endsWith('.func')
+					? collectChunksConsumersInfosFromFuncDirectory(filepath)
+					: collectChunksConsumersInfoRecursively(setup, filepath);
+
+				const dirResults = await dirResultsPromise;
+				dirResults.forEach((chunkConsumers, chunkNumber) => {
+					if (!chunksConsumersInfos.has(chunkNumber)) {
+						chunksConsumersInfos.set(chunkNumber, chunkConsumers);
+					} else {
+						const existingChunkConsumers =
+							chunksConsumersInfos.get(chunkNumber);
+						if (existingChunkConsumers) {
+							for (const consumer of chunkConsumers.values()) {
+								existingChunkConsumers.add(consumer);
+							}
+						}
+					}
+				});
+			}
+		}
+	}
+
+	return chunksConsumersInfos;
+}
+
+async function collectChunksConsumersInfosFromFuncDirectory(
+	filepath: string
+): Promise<ChunksConsumersInfo> {
+	const chunksConsumersInfos: ChunksConsumersInfo = new Map();
+
+	const functionConfig = await readJsonFile<FunctionConfig>(
+		join(filepath, '.vc-config.json')
+	);
+
+	if (!functionConfig) {
```

Comment:
```suggestion
	if (!functionConfig || functionConfig.runtime !== 'edge') {
```

Fixes an issue causing the CLI to crash due to trying to read a non-existent entrypoint field in nodejs runtime function configs.

```
TypeError [ERR_INVALID_ARG_TYPE]: The "path" argument must be of type string. Received undefined
```

Reply:
What does it mean in this case (where the runtime is not marked as "edge"?
Is this actually an error case, since we appear to be trying to process code not marked as edge?

Reply:
> What does it mean in this case (where the runtime is not marked as "edge"? Is this actually an error case, since we appear to be trying to process code not marked as edge?

@petebacondarwin The build output can produce either `edge` or `nodejs` runtime functions, depending on the route. With `nodejs` functions, there are two types; prerender functions and ssr functions. With the first kind, we take the static assets and discard the function, and with the second kind, we log an error during processing the build output. So, whether the runtime not being `edge` is an error case, well, that depends on other factors + checks.

The only type of function code that we use is for `edge` functions, so we should not attempt to process entrypoints for non-edge functions, but returning an error at this point is premature for the reason explained above. Because of where the first part of the deduping logic is added, it occurs before we process all the build output, which means it has no knowledge about which functions we are discarding/ignoring, if we need to error on a function, or what edge functions we've collected. And then the chunk information it collected is used while processing the build output so that we know about the chunks when actually modifying the edge functions.

Reply:
> What does it mean in this case (where the runtime is not marked as "edge"? Is this actually an error case, since we appear to be trying to process code not marked as edge?

Yes it is an actual error case because 'edge' is the only runtime we support as you can see here:
https://github.com/cloudflare/next-on-pages/blob/c0ab05a460ca430a73495f14a39a6cda5c505e1b/packages/next-on-pages/src/buildApplication/generateFunctionsMap.ts#L201-L205

___

I didn't put the check here as I didn't want this function to have more context than necessary to how things work (and not checking the runtime here would just mean that we might be parsing a handful of extra files in failing builds so I thought it'd be ok)

(and I think didn't test it with a broken app 🤦)

Thanks @james-elicx 😄 , I think it is ok to add that check here

Actually one thing to also do would be to see if we can abstract the files processing done both in [processDirectoryRecursively](https://github.com/cloudflare/next-on-pages/blob/c0ab05a460ca430a73495f14a39a6cda5c505e1b/packages/next-on-pages/src/buildApplication/generateFunctionsMap.ts#L128) and in the newly `collectChunksConsumersInfoRecursively` so that we can share the logic and these types of checks.

I'd imagine that can be done separately (it could be very good if we could do it when we have e2es in place as it feels like a bit of a risky refactoring)